<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>The Foundation of BioGA ‚Ä¢ BioGA</title>
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="The Foundation of BioGA">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">BioGA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.99.17</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/BioGA.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/Foundation.html">The Foundation of BioGA</a></li>
    <li><a class="dropdown-item" href="../articles/UsageDemostration.html">Usage Demonstration of BioGA</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/danymukesha/BioGA/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>The Foundation of BioGA</h1>
            <h3 data-toc-skip class="subtitle">Multi-Objective Genetic
Algorithm for Genomic Data Optimization</h3>
                        <h4 data-toc-skip class="author">Dany
Mukesha</h4>
            
            <h4 data-toc-skip class="date">2025-12-04</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/danymukesha/BioGA/blob/main/vignettes/Foundation.Rmd" class="external-link"><code>vignettes/Foundation.Rmd</code></a></small>
      <div class="d-none name"><code>Foundation.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The BioGA utilizes multi-objective Genetic Algorithm (GA)
specifically tailored for the optimization and analysis of genomic data.
A detailed mathematical and theoretical analysis of the algorithms
implemented within this package is described below; covering all major
components of the evolutionary process with formal mathematical
notation, proofs, and a discussion of key properties.</p>
</div>
<div class="section level2">
<h2 id="the-genetic-algorithm-framework">The Genetic Algorithm Framework<a class="anchor" aria-label="anchor" href="#the-genetic-algorithm-framework"></a>
</h2>
<p>The BioGA package employs a standard generational Genetic Algorithm
structure, modified for multi-objective optimization. This framework
involves the iterative evolutionary cycle of initialization, evaluation,
selection, reproduction (crossover and mutation), and replacement.</p>
<div class="section level3">
<h3 id="mathematical-representation">Mathematical Representation<a class="anchor" aria-label="anchor" href="#mathematical-representation"></a>
</h3>
<p>Let us define the core elements of the search space and the
algorithm‚Äôs state:</p>
<ul>
<li>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>G</mi><mo>=</mo><mrow><mo stretchy="true" form="prefix">(</mo><mi>V</mi><mo>,</mo><mi>E</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">G = (V, E)</annotation></semantics></math>
is the gene network, where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>V</mi><mo>=</mo><mo stretchy="false" form="prefix">{</mo><msub><mi>g</mi><mn>1</mn></msub><mo>,</mo><msub><mi>g</mi><mn>2</mn></msub><mo>,</mo><mi>‚Ä¶</mi><mo>,</mo><msub><mi>g</mi><mi>n</mi></msub><mo stretchy="false" form="postfix">}</mo></mrow><annotation encoding="application/x-tex">V = \{g_1, g_2, \ldots, g_n\}</annotation></semantics></math>
are
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>n</mi><annotation encoding="application/x-tex">n</annotation></semantics></math>
genes.</li>
<li>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>X</mi><mo>‚àà</mo><msup><mi>‚Ñù</mi><mrow><mi>n</mi><mo>√ó</mo><mi>m</mi></mrow></msup></mrow><annotation encoding="application/x-tex">X \in \mathbb{R}^{n \times m}</annotation></semantics></math>
is the genomic data matrix, consisting of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>n</mi><annotation encoding="application/x-tex">n</annotation></semantics></math>
genes across
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>m</mi><annotation encoding="application/x-tex">m</annotation></semantics></math>
samples.</li>
<li>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>P</mi><mi>t</mi></msub><mo>‚àà</mo><msup><mi>‚Ñù</mi><mrow><mi>p</mi><mo>√ó</mo><mi>n</mi></mrow></msup></mrow><annotation encoding="application/x-tex">P_t \in \mathbb{R}^{p \times n}</annotation></semantics></math>
is the population matrix at generation
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>t</mi><annotation encoding="application/x-tex">t</annotation></semantics></math>,
containing
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>
individuals (rows) and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>n</mi><annotation encoding="application/x-tex">n</annotation></semantics></math>
gene values (columns).</li>
</ul>
<p>The evolution from one generation to the next,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>P</mi><mi>t</mi></msub><mo>‚Üí</mo><msub><mi>P</mi><mrow><mi>t</mi><mo>+</mo><mn>1</mn></mrow></msub></mrow><annotation encoding="application/x-tex">P_t \to P_{t+1}</annotation></semantics></math>,
is described by the following functional relationship:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>P</mi><mrow><mi>t</mi><mo>+</mo><mn>1</mn></mrow></msub><mo>=</mo><mi>R</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>M</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>C</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>S</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>P</mi><mi>t</mi></msub><mo>,</mo><mi>f</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>P</mi><mi>t</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow><mo stretchy="true" form="postfix">)</mo></mrow><mo>,</mo><mi>X</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo stretchy="true" form="postfix">)</mo></mrow><mo>,</mo><msub><mi>P</mi><mi>t</mi></msub><mo>,</mo><mi>f</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>P</mi><mi>t</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">
P_{t+1} = R(M(C(S(P_t, f(P_t)), X)), P_t, f(P_t))
</annotation></semantics></math></p>
<p>Where the functions represent the core operators:</p>
<ul>
<li>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>f</mi><annotation encoding="application/x-tex">f</annotation></semantics></math>:
<strong>Fitness Evaluation</strong> function.</li>
<li>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>S</mi><annotation encoding="application/x-tex">S</annotation></semantics></math>:
<strong>Selection</strong> operator (NSGA-II inspired).</li>
<li>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>C</mi><annotation encoding="application/x-tex">C</annotation></semantics></math>:
<strong>Crossover</strong> operator (SBX).</li>
<li>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>M</mi><annotation encoding="application/x-tex">M</annotation></semantics></math>:
<strong>Mutation</strong> operator (Adaptive).</li>
<li>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>R</mi><annotation encoding="application/x-tex">R</annotation></semantics></math>:
<strong>Replacement</strong> operator (Elitism + Diversity
Preservation).</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="population-initialization">Population Initialization<a class="anchor" aria-label="anchor" href="#population-initialization"></a>
</h2>
<p>The initialization phase creates the starting population,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>P</mi><mn>0</mn></msub><annotation encoding="application/x-tex">P_0</annotation></semantics></math>,
by drawing random samples directly from the measured genomic data
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>X</mi><annotation encoding="application/x-tex">X</annotation></semantics></math>.</p>
<div class="section level3">
<h3 id="mathematical-formulation">Mathematical Formulation<a class="anchor" aria-label="anchor" href="#mathematical-formulation"></a>
</h3>
<p>Given the genomic data
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>X</mi><mo>‚àà</mo><msup><mi>‚Ñù</mi><mrow><mi>n</mi><mo>√ó</mo><mi>m</mi></mrow></msup></mrow><annotation encoding="application/x-tex">X \in \mathbb{R}^{n \times m}</annotation></semantics></math>
and a population size
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>,
the value of gene
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>
in the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>-th
individual of the initial population is set as:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>P</mi><mn>0</mn></msub><mrow><mo stretchy="true" form="prefix">[</mo><mi>i</mi><mo>,</mo><mi>j</mi><mo stretchy="true" form="postfix">]</mo></mrow><mo>=</mo><mi>X</mi><mrow><mo stretchy="true" form="prefix">[</mo><mi>j</mi><mo>,</mo><mi>k</mi><mo stretchy="true" form="postfix">]</mo></mrow><mspace width="1.0em"></mspace><mtext mathvariant="normal">where</mtext><mspace width="1.0em"></mspace><mi>k</mi><mo>‚àº</mo><mtext mathvariant="normal">Uniform</mtext><mo stretchy="false" form="prefix">{</mo><mn>1</mn><mo>,</mo><mi>‚Ä¶</mi><mo>,</mo><mi>m</mi><mo stretchy="false" form="postfix">}</mo></mrow><annotation encoding="application/x-tex">
P_0[i,j] = X[j, k] \quad \text{where} \quad k \sim \text{Uniform}\{1, \ldots, m\}
</annotation></semantics></math> This means that for each element
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo>,</mo><mi>j</mi><mo stretchy="true" form="postfix">)</mo></mrow><annotation encoding="application/x-tex">(i, j)</annotation></semantics></math>
in
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>P</mi><mn>0</mn></msub><annotation encoding="application/x-tex">P_0</annotation></semantics></math>,
a sample
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>k</mi><annotation encoding="application/x-tex">k</annotation></semantics></math>
is chosen uniformly at random from the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>m</mi><annotation encoding="application/x-tex">m</annotation></semantics></math>
available samples, and the corresponding gene expression value
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>X</mi><mrow><mo stretchy="true" form="prefix">[</mo><mi>j</mi><mo>,</mo><mi>k</mi><mo stretchy="true" form="postfix">]</mo></mrow></mrow><annotation encoding="application/x-tex">X[j, k]</annotation></semantics></math>
is used.</p>
<p><strong>Initialization with Clustering:</strong></p>
<p>If a clustering breakdown is provided (e.g.,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>C</mi><annotation encoding="application/x-tex">C</annotation></semantics></math>
is the set of clusters), the initialization respects the structure:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>P</mi><mn>0</mn></msub><mrow><mo stretchy="true" form="prefix">[</mo><mi>i</mi><mo>,</mo><mi>j</mi><mo stretchy="true" form="postfix">]</mo></mrow><mo>=</mo><mi>X</mi><mrow><mo stretchy="true" form="prefix">[</mo><mi>j</mi><mo>,</mo><mi>k</mi><mo stretchy="true" form="postfix">]</mo></mrow><mspace width="1.0em"></mspace><mo>‚àÄ</mo><mi>j</mi><mo>‚àà</mo><mi>c</mi><mo>,</mo><mspace width="1.0em"></mspace><mi>k</mi><mo>‚àº</mo><mtext mathvariant="normal">Uniform</mtext><mo stretchy="false" form="prefix">{</mo><mn>1</mn><mo>,</mo><mi>‚Ä¶</mi><mo>,</mo><mi>m</mi><mo stretchy="false" form="postfix">}</mo></mrow><annotation encoding="application/x-tex">
P_0[i,j] = X[j, k] \quad \forall j \in c, \quad k \sim \text{Uniform}\{1, \ldots, m\}
</annotation></semantics></math> for all genes
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>
belonging to the same cluster
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>c</mi><annotation encoding="application/x-tex">c</annotation></semantics></math>.</p>
</div>
<div class="section level3">
<h3 id="statistical-properties">Statistical Properties<a class="anchor" aria-label="anchor" href="#statistical-properties"></a>
</h3>
<ol style="list-style-type: decimal">
<li>
<strong>Distribution Preservation:</strong> This method ensures that
the initial population individuals maintain the original statistical
distribution of expression values for each gene
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>.</li>
<li>
<strong>Structural Integrity:</strong> If clustering is used, the
method preserves the pre-defined cluster structure within the initial
population.</li>
<li>
<strong>Expected Value:</strong> The expected value of any element
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>P</mi><mn>0</mn></msub><mrow><mo stretchy="true" form="prefix">[</mo><mi>i</mi><mo>,</mo><mi>j</mi><mo stretchy="true" form="postfix">]</mo></mrow></mrow><annotation encoding="application/x-tex">P_0[i,j]</annotation></semantics></math>
over the initialization process is the mean of the observed expression
for gene
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>ùîº</mi><mrow><mo stretchy="true" form="prefix">[</mo><msub><mi>P</mi><mn>0</mn></msub><mrow><mo stretchy="true" form="prefix">[</mo><mi>i</mi><mo>,</mo><mi>j</mi><mo stretchy="true" form="postfix">]</mo></mrow><mo stretchy="true" form="postfix">]</mo></mrow><mo>=</mo><msub><mi>Œº</mi><mi>j</mi></msub><mo>=</mo><mfrac><mn>1</mn><mi>m</mi></mfrac><munderover><mo>‚àë</mo><mrow><mi>k</mi><mo>=</mo><mn>1</mn></mrow><mi>m</mi></munderover><mi>X</mi><mrow><mo stretchy="true" form="prefix">[</mo><mi>j</mi><mo>,</mo><mi>k</mi><mo stretchy="true" form="postfix">]</mo></mrow></mrow><annotation encoding="application/x-tex">
 \mathbb{E}[P_0[i,j]] = \mu_j = \frac{1}{m}\sum_{k=1}^m X[j,k]
 </annotation></semantics></math>
</li>
</ol>
</div>
</div>
<div class="section level2">
<h2 id="fitness-evaluation">Fitness Evaluation<a class="anchor" aria-label="anchor" href="#fitness-evaluation"></a>
</h2>
<p>The <code>BioGA</code> package utilizes a multi-objective approach,
simultaneously optimizing two conflicting goals: data fidelity
(Expression Difference) and model parsimony (Sparsity).</p>
<div class="section level3">
<h3 id="objective-1-expression-difference-f_1">Objective 1: Expression Difference
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>f</mi><mn>1</mn></msub><annotation encoding="application/x-tex">f_1</annotation></semantics></math>)<a class="anchor" aria-label="anchor" href="#objective-1-expression-difference-f_1"></a>
</h3>
<p>This function measures the squared Euclidean distance between the
individual
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
and all
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>m</mi><annotation encoding="application/x-tex">m</annotation></semantics></math>
observed data samples
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>X</mi><mrow><mi>j</mi><mi>k</mi></mrow></msub><annotation encoding="application/x-tex">X_{jk}</annotation></semantics></math>
for all
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>n</mi><annotation encoding="application/x-tex">n</annotation></semantics></math>
genes. It quantifies how accurately the individual reflects the overall
expression patterns.</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>f</mi><mn>1</mn></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><munderover><mo>‚àë</mo><mrow><mi>j</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><munderover><mo>‚àë</mo><mrow><mi>k</mi><mo>=</mo><mn>1</mn></mrow><mi>m</mi></munderover><msup><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>X</mi><mrow><mi>j</mi><mi>k</mi></mrow></msub><mo>‚àí</mo><msub><mi>P</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow><mn>2</mn></msup></mrow><annotation encoding="application/x-tex">
f_1(i) = \sum_{j=1}^n \sum_{k=1}^m (X_{jk} - P_{ij})^2
</annotation></semantics></math></p>
<div class="section level4">
<h4 id="properties-of-f_1">Properties of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>f</mi><mn>1</mn></msub><annotation encoding="application/x-tex">f_1</annotation></semantics></math><a class="anchor" aria-label="anchor" href="#properties-of-f_1"></a>
</h4>
<ol style="list-style-type: decimal">
<li>
<strong>Convexity:</strong>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>f</mi><mn>1</mn></msub><annotation encoding="application/x-tex">f_1</annotation></semantics></math>
is a convex function, which is a desirable property for optimization as
it guarantees that any local minimum is also a global minimum in the
absence of other constraints. The minimum occurs when
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>P</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">P_{ij}</annotation></semantics></math>
is the mean expression
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Œº</mi><mi>j</mi></msub><annotation encoding="application/x-tex">\mu_j</annotation></semantics></math>.</li>
<li>
<strong>Gradient (for theoretical analysis):</strong> The partial
derivative with respect to the gene value
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>P</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">P_{ij}</annotation></semantics></math>
is:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>‚àá</mi><msub><mi>P</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub></msub><msub><mi>f</mi><mn>1</mn></msub><mo>=</mo><mo>‚àí</mo><mn>2</mn><munderover><mo>‚àë</mo><mrow><mi>k</mi><mo>=</mo><mn>1</mn></mrow><mi>m</mi></munderover><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>X</mi><mrow><mi>j</mi><mi>k</mi></mrow></msub><mo>‚àí</mo><msub><mi>P</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">
 \nabla_{P_{ij}} f_1 = -2\sum_{k=1}^m (X_{jk} - P_{ij})
 </annotation></semantics></math>
</li>
</ol>
</div>
</div>
<div class="section level3">
<h3 id="objective-2-sparsity-f_2">Objective 2: Sparsity
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>f</mi><mn>2</mn></msub><annotation encoding="application/x-tex">f_2</annotation></semantics></math>)<a class="anchor" aria-label="anchor" href="#objective-2-sparsity-f_2"></a>
</h3>
<p>This objective promotes parsimonious solutions by penalizing non-zero
gene values. It measures the fraction of non-zero genes within an
individual
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>.</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>f</mi><mn>2</mn></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><mfrac><mrow><munderover><mo>‚àë</mo><mrow><mi>j</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><mi>I</mi><mrow><mo stretchy="true" form="prefix">(</mo><mrow><mo stretchy="true" form="prefix">|</mo><msub><mi>P</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo stretchy="true" form="postfix">|</mo></mrow><mo>&gt;</mo><mi>œµ</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><mi>n</mi></mfrac></mrow><annotation encoding="application/x-tex">
f_2(i) = \frac{\sum_{j=1}^n I(|P_{ij}| &gt; \epsilon)}{n}
</annotation></semantics></math></p>
<p>Here,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>I</mi><annotation encoding="application/x-tex">I</annotation></semantics></math>
is the indicator function, which equals 1 if the condition is met, and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>œµ</mi><annotation encoding="application/x-tex">\epsilon</annotation></semantics></math>
is a small constant (typically
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mn>10</mn><mrow><mo>‚àí</mo><mn>6</mn></mrow></msup><annotation encoding="application/x-tex">10^{-6}</annotation></semantics></math>)
used to define ‚Äúnon-zero.‚Äù</p>
<div class="section level4">
<h4 id="properties-of-f_2">Properties of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>f</mi><mn>2</mn></msub><annotation encoding="application/x-tex">f_2</annotation></semantics></math><a class="anchor" aria-label="anchor" href="#properties-of-f_2"></a>
</h4>
<ol style="list-style-type: decimal">
<li>
<strong>Mathematical Form:</strong>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>f</mi><mn>2</mn></msub><annotation encoding="application/x-tex">f_2</annotation></semantics></math>
is inherently <strong>Non-convex</strong> and
<strong>Non-differentiable</strong> due to the indicator function
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>I</mi><annotation encoding="application/x-tex">I</annotation></semantics></math>.
This makes standard gradient-based optimization methods unsuitable,
justifying the use of a heuristic approach like the GA.</li>
<li>
<strong>Goal:</strong> It actively encourages sparse solutions,
where many gene values are effectively zero.</li>
<li>
<strong>Range:</strong> The range of the function is restricted to
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo stretchy="true" form="prefix">[</mo><mn>0</mn><mo>,</mo><mn>1</mn><mo stretchy="true" form="postfix">]</mo></mrow><annotation encoding="application/x-tex">[0, 1]</annotation></semantics></math>,
where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>0</mn><annotation encoding="application/x-tex">0</annotation></semantics></math>
represents a fully sparse individual
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>P</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>=</mo><mn>0</mn></mrow><annotation encoding="application/x-tex">P_{ij} = 0</annotation></semantics></math>
for all
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>)
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>1</mn><annotation encoding="application/x-tex">1</annotation></semantics></math>
represents a non-sparse individual.</li>
</ol>
</div>
</div>
<div class="section level3">
<h3 id="combined-fitness">Combined Fitness<a class="anchor" aria-label="anchor" href="#combined-fitness"></a>
</h3>
<p>The overall fitness value
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>F</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">F(i)</annotation></semantics></math>
for an individual
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
is a weighted combination of the two objectives:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>F</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><msub><mi>w</mi><mn>1</mn></msub><msub><mi>f</mi><mn>1</mn></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>+</mo><msub><mi>w</mi><mn>2</mn></msub><msub><mi>f</mi><mn>2</mn></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">
F(i) = w_1f_1(i) + w_2f_2(i)
</annotation></semantics></math></p>
<p>Where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>w</mi><mn>1</mn></msub><annotation encoding="application/x-tex">w_1</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>w</mi><mn>2</mn></msub><annotation encoding="application/x-tex">w_2</annotation></semantics></math>
are user-specified weights that dictate the trade-off between minimizing
expression error
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>f</mi><mn>1</mn></msub><annotation encoding="application/x-tex">f_1</annotation></semantics></math>)
and maximizing sparsity
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>f</mi><mn>2</mn></msub><annotation encoding="application/x-tex">f_2</annotation></semantics></math>).</p>
</div>
</div>
<div class="section level2">
<h2 id="selection-nsga-ii-inspired">Selection (NSGA-II Inspired)<a class="anchor" aria-label="anchor" href="#selection-nsga-ii-inspired"></a>
</h2>
<p>The selection mechanism is based on the principles of the
Non-dominated Sorting Genetic Algorithm II (NSGA-II), ensuring that
evolutionary pressure drives the population toward the Pareto front.</p>
<div class="section level3">
<h3 id="domination-criteria">Domination Criteria<a class="anchor" aria-label="anchor" href="#domination-criteria"></a>
</h3>
<p>Individual
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
dominates individual
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>
(denoted
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>i</mi><mo>‚âª</mo><mi>j</mi></mrow><annotation encoding="application/x-tex">i \succ j</annotation></semantics></math>)
if and only if:</p>
<ol style="list-style-type: decimal">
<li>Individual
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
is no worse than
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>
on all objectives, and</li>
<li>Individual
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
is strictly better than
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>
on at least one objective.</li>
</ol>
<p>Formally:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>i</mi><mo>‚âª</mo><mi>j</mi><mspace width="1.0em"></mspace><mtext mathvariant="normal">iff</mtext><mspace width="1.0em"></mspace><mo>‚àÄ</mo><mi>k</mi><mo>‚àà</mo><mo stretchy="false" form="prefix">{</mo><mn>1</mn><mo>,</mo><mn>2</mn><mo stretchy="false" form="postfix">}</mo><mo>:</mo><msub><mi>f</mi><mi>k</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>‚â§</mo><msub><mi>f</mi><mi>k</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>j</mi><mo stretchy="true" form="postfix">)</mo></mrow><mspace width="1.0em"></mspace><mtext mathvariant="normal">and</mtext><mspace width="1.0em"></mspace><mo>‚àÉ</mo><mi>k</mi><mo>‚àà</mo><mo stretchy="false" form="prefix">{</mo><mn>1</mn><mo>,</mo><mn>2</mn><mo stretchy="false" form="postfix">}</mo><mo>:</mo><msub><mi>f</mi><mi>k</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>&lt;</mo><msub><mi>f</mi><mi>k</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>j</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">
i \succ j \quad \text{iff} \quad \forall k \in \{1, 2\}: f_k(i) \leq f_k(j) \quad \text{and} \quad \exists k \in \{1, 2\}: f_k(i) &lt; f_k(j)
</annotation></semantics></math></p>
<div class="section level4">
<h4 id="proof-of-partial-order">Proof of Partial Order<a class="anchor" aria-label="anchor" href="#proof-of-partial-order"></a>
</h4>
<p>The two-objective domination relation
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo stretchy="true" form="prefix">(</mo><mo>‚âª</mo><mo stretchy="true" form="postfix">)</mo></mrow><annotation encoding="application/x-tex">(\succ)</annotation></semantics></math>
defines a <strong>partial order</strong> on the solution set because it
satisfies the following properties:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Irreflexive:</strong> No individual dominates itself. If
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>i</mi><mo>=</mo><mi>j</mi></mrow><annotation encoding="application/x-tex">i = j</annotation></semantics></math>,
the second condition
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo>‚àÉ</mo><mi>k</mi><mo>:</mo><msub><mi>f</mi><mi>k</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>&lt;</mo><msub><mi>f</mi><mi>k</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>j</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\exists k: f_k(i) &lt; f_k(j)</annotation></semantics></math>)
is false.</li>
<li>
<strong>Antisymmetric Property:</strong> If
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>i</mi><mo>‚âª</mo><mi>j</mi></mrow><annotation encoding="application/x-tex">i \succ j</annotation></semantics></math>,
then
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>
cannot dominate
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>.
If
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
is strictly better on at least one objective,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>
must be strictly worse
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo>‚àÄ</mo><mi>k</mi><mo>:</mo><msub><mi>f</mi><mi>k</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>j</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>‚â∞</mo><msub><mi>f</mi><mi>k</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\forall k: f_k(j) \not\leq f_k(i)</annotation></semantics></math>).</li>
<li>
<strong>Transitive:</strong> If
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>i</mi><mo>‚âª</mo><mi>j</mi></mrow><annotation encoding="application/x-tex">i \succ j</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>j</mi><mo>‚âª</mo><mi>k</mi></mrow><annotation encoding="application/x-tex">j \succ k</annotation></semantics></math>,
then
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>i</mi><mo>‚âª</mo><mi>k</mi></mrow><annotation encoding="application/x-tex">i \succ k</annotation></semantics></math>.
This follows directly from the transitivity of the combined
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mo>‚â§</mo><annotation encoding="application/x-tex">\leq</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mo>&lt;</mo><annotation encoding="application/x-tex">&lt;</annotation></semantics></math>
relations.</li>
</ol>
</div>
</div>
<div class="section level3">
<h3 id="front-construction">Front Construction<a class="anchor" aria-label="anchor" href="#front-construction"></a>
</h3>
<p>The population is sorted into non-dominated fronts:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Domination Metrics:</strong> For every individual, we
compute its <strong>domination count</strong> (number of individuals
that dominate it) and its <strong>dominated set</strong> (set of
individuals it dominates).</li>
<li>
<strong>First Front
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>‚Ñ±</mi><mn>1</mn></msub><annotation encoding="application/x-tex">\mathcal{F}_1</annotation></semantics></math>):</strong>
This front consists of all individuals whose domination count is
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>0</mn><annotation encoding="application/x-tex">0</annotation></semantics></math>.
These are the current non-dominated solutions.</li>
<li>
<strong>Subsequent Fronts:</strong> Individuals in
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>‚Ñ±</mi><mi>t</mi></msub><annotation encoding="application/x-tex">\mathcal{F}_t</annotation></semantics></math>
are temporarily removed. The domination counts of the individuals they
dominated are reduced by one. The next front,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>‚Ñ±</mi><mrow><mi>t</mi><mo>+</mo><mn>1</mn></mrow></msub><annotation encoding="application/x-tex">\mathcal{F}_{t+1}</annotation></semantics></math>,
consists of all remaining individuals whose updated domination count is
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>0</mn><annotation encoding="application/x-tex">0</annotation></semantics></math>.
This process continues until the entire population is sorted.</li>
</ol>
<div class="section level4">
<h4 id="theorem-on-computational-complexity">Theorem on Computational Complexity<a class="anchor" aria-label="anchor" href="#theorem-on-computational-complexity"></a>
</h4>
<p>The non-dominated sorting and front construction algorithm terminates
in quadratic time complexity relative to the population size.</p>
<p><strong>Theorem:</strong> The front construction algorithm terminates
in
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>O</mi><mrow><mo stretchy="true" form="prefix">(</mo><msup><mi>p</mi><mn>2</mn></msup><mi>o</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">O(p^2o)</annotation></semantics></math>
time, where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>
is the population size and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>o</mi><annotation encoding="application/x-tex">o</annotation></semantics></math>
is the number of objectives.</p>
<p><strong>Proof Sketch:</strong></p>
<ul>
<li>The pairwise domination check between any two individuals is an
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>O</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>o</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">O(o)</annotation></semantics></math>
operation.</li>
<li>The computation of the domination count and dominated set for all
individuals requires
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>O</mi><mrow><mo stretchy="true" form="prefix">(</mo><msup><mi>p</mi><mn>2</mn></msup><mi>o</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">O(p^2o)</annotation></semantics></math>
time, as it involves
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo stretchy="true" form="prefix">(</mo><mfrac linethickness="0"><mi>p</mi><mn>2</mn></mfrac><mo stretchy="true" form="postfix">)</mo></mrow><annotation encoding="application/x-tex">\binom{p}{2}</annotation></semantics></math>
comparisons.</li>
<li>Once these metrics are computed, the iterative process of assigning
individuals to fronts requires
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>O</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>p</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">O(p)</annotation></semantics></math>
time per front, leading to an overall complexity dominated by the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>O</mi><mrow><mo stretchy="true" form="prefix">(</mo><msup><mi>p</mi><mn>2</mn></msup><mi>o</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">O(p^2o)</annotation></semantics></math>
initial comparison phase.</li>
</ul>
</div>
</div>
</div>
<div class="section level2">
<h2 id="crossover-simulated-binary-crossover---sbx">Crossover (Simulated Binary Crossover - SBX)<a class="anchor" aria-label="anchor" href="#crossover-simulated-binary-crossover---sbx"></a>
</h2>
<p>The Simulated Binary Crossover (SBX) is a prominent operator for
real-coded GAs, designed to mimic the convergence properties of binary
crossover while operating directly on continuous variables.</p>
<p>Given two parent individuals,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>x</mi><mo>,</mo><mi>y</mi><mo>‚àà</mo><msup><mi>‚Ñù</mi><mi>n</mi></msup></mrow><annotation encoding="application/x-tex">x, y \in \mathbb{R}^n</annotation></semantics></math>,
the SBX operator generates a new offspring vector
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>z</mi><annotation encoding="application/x-tex">z</annotation></semantics></math>.
For each gene
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>:</p>
<p>With crossover probability
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>p</mi><mi>c</mi></msub><annotation encoding="application/x-tex">p_c</annotation></semantics></math>,
the offspring gene
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>z</mi><mi>j</mi></msub><annotation encoding="application/x-tex">z_j</annotation></semantics></math>
is determined: 1. A random number
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>u</mi><annotation encoding="application/x-tex">u</annotation></semantics></math>
is drawn:
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>u</mi><mo>‚àº</mo><mtext mathvariant="normal">Uniform</mtext><mrow><mo stretchy="true" form="prefix">(</mo><mn>0</mn><mo>,</mo><mn>1</mn><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">u \sim \text{Uniform}(0,1)</annotation></semantics></math>.
2. The spreading factor
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Œ≤</mi><annotation encoding="application/x-tex">\beta</annotation></semantics></math>
is computed based on the distribution index
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Œ∑</mi><annotation encoding="application/x-tex">\eta</annotation></semantics></math>:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Œ≤</mi><mo>=</mo><mrow><mo stretchy="true" form="prefix">{</mo><mtable><mtr><mtd columnalign="left" style="text-align: left"><msup><mrow><mo stretchy="true" form="prefix">(</mo><mn>2</mn><mi>u</mi><mo stretchy="true" form="postfix">)</mo></mrow><mrow><mn>1</mn><mi>/</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>Œ∑</mi><mo>+</mo><mn>1</mn><mo stretchy="true" form="postfix">)</mo></mrow></mrow></msup></mtd><mtd columnalign="left" style="text-align: left"><mrow><mtext mathvariant="normal">if </mtext><mspace width="0.333em"></mspace></mrow><mi>u</mi><mo>‚â§</mo><mn>0.5</mn></mtd></mtr><mtr><mtd columnalign="left" style="text-align: left"><msup><mrow><mo stretchy="true" form="prefix">(</mo><mfrac><mn>1</mn><mrow><mn>2</mn><mrow><mo stretchy="true" form="prefix">(</mo><mn>1</mn><mo>‚àí</mo><mi>u</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow></mfrac><mo stretchy="true" form="postfix">)</mo></mrow><mrow><mn>1</mn><mi>/</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>Œ∑</mi><mo>+</mo><mn>1</mn><mo stretchy="true" form="postfix">)</mo></mrow></mrow></msup></mtd><mtd columnalign="left" style="text-align: left"><mtext mathvariant="normal">otherwise</mtext></mtd></mtr></mtable></mrow></mrow><annotation encoding="application/x-tex">
    \beta = \begin{cases}
    (2u)^{1/(\eta+1)} &amp; \text{if } u \leq 0.5 \\
    \left(\frac{1}{2(1-u)}\right)^{1/(\eta+1)} &amp; \text{otherwise}
    \end{cases}
    </annotation></semantics></math> 3. The gene value
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>z</mi><mi>j</mi></msub><annotation encoding="application/x-tex">z_j</annotation></semantics></math>
is calculated:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>z</mi><mi>j</mi></msub><mo>=</mo><mn>0.5</mn><mrow><mo stretchy="true" form="prefix">[</mo><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>x</mi><mi>j</mi></msub><mo>+</mo><msub><mi>y</mi><mi>j</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow><mo>‚àí</mo><mi>Œ≤</mi><mrow><mo stretchy="true" form="prefix">|</mo><msub><mi>y</mi><mi>j</mi></msub><mo>‚àí</mo><msub><mi>x</mi><mi>j</mi></msub><mo stretchy="true" form="postfix">|</mo></mrow><mo stretchy="true" form="postfix">]</mo></mrow></mrow><annotation encoding="application/x-tex">
    z_j = 0.5[(x_j + y_j) - \beta|y_j - x_j|]
    </annotation></semantics></math></p>
<p>Otherwise (with probability
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>1</mn><mo>‚àí</mo><msub><mi>p</mi><mi>c</mi></msub></mrow><annotation encoding="application/x-tex">1-p_c</annotation></semantics></math>):
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>z</mi><mi>j</mi></msub><mo>=</mo><msub><mi>x</mi><mi>j</mi></msub><mspace width="1.0em"></mspace><mtext mathvariant="normal">or</mtext><mspace width="1.0em"></mspace><msub><mi>z</mi><mi>j</mi></msub><mo>=</mo><msub><mi>y</mi><mi>j</mi></msub></mrow><annotation encoding="application/x-tex">
z_j = x_j \quad \text{or} \quad z_j = y_j
</annotation></semantics></math></p>
<div class="section level3">
<h3 id="properties-of-sbx">Properties of SBX<a class="anchor" aria-label="anchor" href="#properties-of-sbx"></a>
</h3>
<ol style="list-style-type: decimal">
<li>
<strong>Mean Preservation:</strong> The expected value of the
offspring gene is average of the parents‚Äô genes:
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>ùîº</mi><mrow><mo stretchy="true" form="prefix">[</mo><msub><mi>z</mi><mi>j</mi></msub><mo stretchy="true" form="postfix">]</mo></mrow><mo>=</mo><mfrac><mrow><msub><mi>x</mi><mi>j</mi></msub><mo>+</mo><msub><mi>y</mi><mi>j</mi></msub></mrow><mn>2</mn></mfrac></mrow><annotation encoding="application/x-tex">\mathbb{E}[z_j] = \frac{x_j + y_j}{2}</annotation></semantics></math>.</li>
<li>
<strong>Distribution Control:</strong> The distribution index
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Œ∑</mi><annotation encoding="application/x-tex">\eta</annotation></semantics></math>
controls the proximity of the offspring to the parents.</li>
<li>
<strong>Low
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Œ∑</mi><annotation encoding="application/x-tex">\eta</annotation></semantics></math>
(e.g.,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Œ∑</mi><mo>‚Üí</mo><mn>0</mn></mrow><annotation encoding="application/x-tex">\eta \to 0</annotation></semantics></math>):</strong>
Produces a wide spread and approaches uniform crossover, generating many
extreme values.</li>
<li>
<strong>High
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Œ∑</mi><annotation encoding="application/x-tex">\eta</annotation></semantics></math>
(e.g.,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Œ∑</mi><mo>‚Üí</mo><mi>‚àû</mi></mrow><annotation encoding="application/x-tex">\eta \to \infty</annotation></semantics></math>):</strong>
Clusters offspring tightly around the parents, approaching no
crossover.</li>
</ol>
</div>
</div>
<div class="section level2">
<h2 id="mutation">Mutation<a class="anchor" aria-label="anchor" href="#mutation"></a>
</h2>
<p>The mutation operation introduces random small changes to the
offspring for exploration. The <code>BioGA</code> uses an adaptive
mutation rate and incorporates network constraints.</p>
<p>For each gene
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>,
with adaptive probability
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>p</mi><mi>m</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>t</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">p_m(t)</annotation></semantics></math>:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>p</mi><mi>m</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>t</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><msub><mi>p</mi><mn>0</mn></msub><mrow><mo stretchy="true" form="prefix">(</mo><mn>1</mn><mo>+</mo><mn>0.5</mn><mfrac><mi>t</mi><mi>T</mi></mfrac><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">
p_m(t) = p_0\left(1 + 0.5\frac{t}{T}\right)
</annotation></semantics></math> where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>p</mi><mn>0</mn></msub><annotation encoding="application/x-tex">p_0</annotation></semantics></math>
is the initial rate,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>t</mi><annotation encoding="application/x-tex">t</annotation></semantics></math>
is the current generation, and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>T</mi><annotation encoding="application/x-tex">T</annotation></semantics></math>
is the total number of generations. This rate increases linearly over
time to enhance exploration in later stages.</p>
<p>The change in the gene value,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>z</mi><mi>j</mi></msub><annotation encoding="application/x-tex">z_j</annotation></semantics></math>,
is applied as follows:</p>
<ol style="list-style-type: decimal">
<li><p>A Gaussian perturbation is generated:
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>Œî</mi><mi>j</mi></msub><mo>‚àº</mo><mi>N</mi><mrow><mo stretchy="true" form="prefix">(</mo><mn>0</mn><mo>,</mo><msup><mi>œÉ</mi><mn>2</mn></msup><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\Delta_j \sim N(0, \sigma^2)</annotation></semantics></math>.</p></li>
<li><p><strong>If a Gene Network
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>N</mi><annotation encoding="application/x-tex">N</annotation></semantics></math>
is provided
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>N</mi><mrow><mi>j</mi><mi>k</mi></mrow></msub><annotation encoding="application/x-tex">N_{jk}</annotation></semantics></math>
being the adjacency matrix):</strong> The mutation is penalized by the
network connectivity:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>z</mi><mi>j</mi></msub><mo>‚Üê</mo><msub><mi>z</mi><mi>j</mi></msub><mo>+</mo><msub><mi>Œî</mi><mi>j</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><mn>1</mn><mo>‚àí</mo><munder><mo>‚àë</mo><mi>k</mi></munder><msub><mi>N</mi><mrow><mi>j</mi><mi>k</mi></mrow></msub><msub><mi>z</mi><mi>k</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">
 z_j \leftarrow z_j + \Delta_j\left(1 - \sum_{k} N_{jk}z_k\right)
 </annotation></semantics></math></p></li>
<li><p><strong>If no Network is provided:</strong> Standard Gaussian
mutation is applied:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>z</mi><mi>j</mi></msub><mo>‚Üê</mo><msub><mi>z</mi><mi>j</mi></msub><mo>+</mo><msub><mi>Œî</mi><mi>j</mi></msub></mrow><annotation encoding="application/x-tex">
 z_j \leftarrow z_j + \Delta_j
 </annotation></semantics></math></p></li>
</ol>
<div class="section level3">
<h3 id="properties-of-adaptive-network-mutation">Properties of Adaptive Network Mutation<a class="anchor" aria-label="anchor" href="#properties-of-adaptive-network-mutation"></a>
</h3>
<ol style="list-style-type: decimal">
<li>
<strong>Adaptive Rate:</strong> The mutation rate monotonically
increases with generation
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>t</mi><annotation encoding="application/x-tex">t</annotation></semantics></math>,
preventing premature convergence and maintaining diversity late in the
run.</li>
<li>
<strong>Network Constraint Impact:</strong> The term
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo stretchy="true" form="prefix">(</mo><mn>1</mn><mo>‚àí</mo><msub><mo>‚àë</mo><mi>k</mi></msub><msub><mi>N</mi><mrow><mi>j</mi><mi>k</mi></mrow></msub><msub><mi>z</mi><mi>k</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow><annotation encoding="application/x-tex">(1 - \sum_{k} N_{jk}z_k)</annotation></semantics></math>
acts as a coefficient. Highly connected genes (large
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mo>‚àë</mo><mi>k</mi></msub><msub><mi>N</mi><mrow><mi>j</mi><mi>k</mi></mrow></msub></mrow><annotation encoding="application/x-tex">\sum_{k} N_{jk}</annotation></semantics></math>)
experience a reduced magnitude of mutation, preserving known network
structures.</li>
<li>
<strong>Expected Change:</strong> Since
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>ùîº</mi><mrow><mo stretchy="true" form="prefix">[</mo><msub><mi>Œî</mi><mi>j</mi></msub><mo stretchy="true" form="postfix">]</mo></mrow><mo>=</mo><mn>0</mn></mrow><annotation encoding="application/x-tex">\mathbb{E}[\Delta_j] = 0</annotation></semantics></math>,
the expected change remains zero:
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>ùîº</mi><mrow><mo stretchy="true" form="prefix">[</mo><mi>Œî</mi><msub><mi>z</mi><mi>j</mi></msub><mo stretchy="true" form="postfix">]</mo></mrow><mo>=</mo><mn>0</mn></mrow><annotation encoding="application/x-tex">\mathbb{E}[\Delta z_j] = 0</annotation></semantics></math>.</li>
<li>
<strong>Variance Control:</strong> For the network-constrained case,
the variance of the change is:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mtext mathvariant="normal">Var</mtext><mrow><mo stretchy="true" form="prefix">(</mo><mi>Œî</mi><msub><mi>z</mi><mi>j</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><msup><mi>œÉ</mi><mn>2</mn></msup><msup><mrow><mo stretchy="true" form="prefix">(</mo><mn>1</mn><mo>‚àí</mo><munder><mo>‚àë</mo><mi>k</mi></munder><msub><mi>N</mi><mrow><mi>j</mi><mi>k</mi></mrow></msub><msub><mi>z</mi><mi>k</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow><mn>2</mn></msup></mrow><annotation encoding="application/x-tex">
 \text{Var}(\Delta z_j) = \sigma^2\left(1 - \sum_{k} N_{jk}z_k\right)^2
 </annotation></semantics></math> This mathematical structure formalizes
how network relationships dynamically control evolutionary
exploration.</li>
</ol>
</div>
</div>
<div class="section level2">
<h2 id="replacement">Replacement<a class="anchor" aria-label="anchor" href="#replacement"></a>
</h2>
<p>The replacement operator
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>R</mi><annotation encoding="application/x-tex">R</annotation></semantics></math>
constructs the next population
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>P</mi><mrow><mi>t</mi><mo>+</mo><mn>1</mn></mrow></msub><annotation encoding="application/x-tex">P_{t+1}</annotation></semantics></math>
from the current population
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>P</mi><mi>t</mi></msub><annotation encoding="application/x-tex">P_t</annotation></semantics></math>
and the offspring
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>O</mi><mi>t</mi></msub><annotation encoding="application/x-tex">O_t</annotation></semantics></math>.
It employs an Elitism strategy combined with a metric for diversity
preservation.</p>
<ol style="list-style-type: decimal">
<li>
<strong>Elitism:</strong> The best individual,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>x</mi><mo>*</mo></msup><annotation encoding="application/x-tex">x^*</annotation></semantics></math>,
defined as the solution minimizing the primary objective
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>f</mi><mn>1</mn></msub><annotation encoding="application/x-tex">f_1</annotation></semantics></math>,
is always retained in the next generation:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msup><mi>x</mi><mo>*</mo></msup><mo>=</mo><msub><mtext mathvariant="normal">argmin</mtext><mrow><mi>x</mi><mo>‚àà</mo><msub><mi>P</mi><mi>t</mi></msub><mo>‚à™</mo><msub><mi>O</mi><mi>t</mi></msub></mrow></msub><msub><mi>f</mi><mn>1</mn></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>x</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">
 x^* = \text{argmin}_{x \in P_t \cup O_t} f_1(x)
 </annotation></semantics></math>
</li>
<li>
<strong>Diversity-Preserving Replacement:</strong> For the remaining
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>p</mi><mo>‚àí</mo><mn>1</mn></mrow><annotation encoding="application/x-tex">p-1</annotation></semantics></math>
slots in
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>P</mi><mrow><mi>t</mi><mo>+</mo><mn>1</mn></mrow></msub><annotation encoding="application/x-tex">P_{t+1}</annotation></semantics></math>,
a non-elitist replacement strategy is performed iteratively:
<ul>
<li>Randomly select an individual
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>x</mi><annotation encoding="application/x-tex">x</annotation></semantics></math>
from the current population
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>P</mi><mi>t</mi></msub><annotation encoding="application/x-tex">P_t</annotation></semantics></math>.</li>
<li>Select an offspring
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>y</mi><annotation encoding="application/x-tex">y</annotation></semantics></math>
from the newly generated offspring
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>O</mi><mi>t</mi></msub><annotation encoding="application/x-tex">O_t</annotation></semantics></math>.</li>
<li>Replace
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>x</mi><annotation encoding="application/x-tex">x</annotation></semantics></math>
with
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>y</mi><annotation encoding="application/x-tex">y</annotation></semantics></math>
only if the increase in diversity is significant:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mrow><mtext mathvariant="normal">Replace </mtext><mspace width="0.333em"></mspace></mrow><mi>x</mi><mo>‚Üê</mo><mi>y</mi><mspace width="1.0em"></mspace><mtext mathvariant="normal">if</mtext><mspace width="1.0em"></mspace><mtext mathvariant="normal">diversity</mtext><mrow><mo stretchy="true" form="prefix">(</mo><mi>x</mi><mo>,</mo><mi>y</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>&gt;</mo><mi>œµ</mi></mrow><annotation encoding="application/x-tex">
  \text{Replace } x \leftarrow y \quad \text{if} \quad \text{diversity}(x,y) &gt; \epsilon
  </annotation></semantics></math> Where the diversity metric is the
squared Euclidean distance:
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mtext mathvariant="normal">diversity</mtext><mrow><mo stretchy="true" form="prefix">(</mo><mi>x</mi><mo>,</mo><mi>y</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><mo stretchy="false" form="postfix">‚à•</mo><mi>x</mi><mo>‚àí</mo><mi>y</mi><msubsup><mo stretchy="false" form="postfix">‚à•</mo><mn>2</mn><mn>2</mn></msubsup></mrow><annotation encoding="application/x-tex">\text{diversity}(x,y) = \|x - y\|_2^2</annotation></semantics></math>.</li>
</ul>
</li>
</ol>
<div class="section level3">
<h3 id="theorem-on-population-quality">Theorem on Population Quality<a class="anchor" aria-label="anchor" href="#theorem-on-population-quality"></a>
</h3>
<p><strong>Theorem:</strong> This replacement strategy preserves elitism
while ensuring that the population‚Äôs diversity does not collapse into a
single point.</p>
<p><strong>Proof Sketch:</strong></p>
<ol style="list-style-type: decimal">
<li>
<strong>Elitism Preservation (Quality):</strong> By explicitly
retaining
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>x</mi><mo>*</mo></msup><annotation encoding="application/x-tex">x^*</annotation></semantics></math>,
the search process ensures that the best solution found so far across
the primary objective
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>f</mi><mn>1</mn></msub><annotation encoding="application/x-tex">f_1</annotation></semantics></math>
is never lost. This guarantees monotonic, non-decreasing convergence
quality on
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>f</mi><mn>1</mn></msub><annotation encoding="application/x-tex">f_1</annotation></semantics></math>.</li>
<li>
<strong>Diversity Maintenance:</strong> Replacements are conditional
on
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mtext mathvariant="normal">diversity</mtext><mrow><mo stretchy="true" form="prefix">(</mo><mi>x</mi><mo>,</mo><mi>y</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>&gt;</mo><mi>œµ</mi></mrow><annotation encoding="application/x-tex">\text{diversity}(x,y) &gt; \epsilon</annotation></semantics></math>.
Since only replacements that increase or maintain the distance (and
hence diversity) above a threshold are accepted, the expected diversity
of the population is non-decreasing over the generations, effectively
counteracting sampling drift.</li>
</ol>
</div>
</div>
<div class="section level2">
<h2 id="convergence-analysis">Convergence Analysis<a class="anchor" aria-label="anchor" href="#convergence-analysis"></a>
</h2>
<p>The convergence of the BioGA algorithm, derived from established
Multi-objective Evolutionary Algorithm (MOEA) theory, guarantees
convergence to the optimal set under general conditions.</p>
<div class="section level3">
<h3 id="assumptions">Assumptions<a class="anchor" aria-label="anchor" href="#assumptions"></a>
</h3>
<p>For convergence proofs in MOEAs, the following are often assumed:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Finite Search Space:</strong> While the gene values are
continuous, they can be discretized in theory (or bounded in
practice).</li>
<li>
<strong>Strictly Positive Mutation Probability:</strong>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>p</mi><mi>m</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>t</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>&gt;</mo><mn>0</mn></mrow><annotation encoding="application/x-tex">p_m(t) &gt; 0</annotation></semantics></math>
for all
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>t</mi><annotation encoding="application/x-tex">t</annotation></semantics></math>,
which ensures ergodicity.</li>
<li>
<strong>Elitism:</strong> The best solutions are preserved across
generations.</li>
</ol>
</div>
<div class="section level3">
<h3 id="theorem-on-convergence">Theorem on Convergence<a class="anchor" aria-label="anchor" href="#theorem-on-convergence"></a>
</h3>
<p><strong>Theorem:</strong> Under the assumptions above, the BioGA
algorithm converges in probability to the global Pareto front of the
multi-objective optimization problem.</p>
<p><strong>Proof Sketch:</strong></p>
<ol style="list-style-type: decimal">
<li>
<strong>Preservation of Optimality:</strong> The NSGA-II selection
and Elitism replacement ensure that once a solution is found on the
Pareto front, it or a strictly better non-dominated alternative is
retained.</li>
<li>
<strong>Ergodicity:</strong> The strictly positive mutation
probability guarantees that any point in the search space can
theoretically be reached from any other point in a finite number of
steps.</li>
<li>
<strong>Convergence Result:</strong> Based on general MOEA
convergence theorems (e.g., Rudolph 1998, or similar proofs for
NSGA-II), the combination of selection pressure toward non-dominated
solutions and exploratory power (mutation/crossover) ensures that the
population will tend toward the true Pareto optimal set as
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>t</mi><mo>‚Üí</mo><mi>‚àû</mi></mrow><annotation encoding="application/x-tex">t \to \infty</annotation></semantics></math>.</li>
</ol>
</div>
</div>
<div class="section level2">
<h2 id="computational-complexity">Computational Complexity<a class="anchor" aria-label="anchor" href="#computational-complexity"></a>
</h2>
<p>Analyzing the complexity is crucial for understanding the algorithm‚Äôs
performance on large genomic datasets. Key variables are defined:</p>
<ul>
<li>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>
= population size</li>
<li>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>n</mi><annotation encoding="application/x-tex">n</annotation></semantics></math>
= number of genes</li>
<li>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>m</mi><annotation encoding="application/x-tex">m</annotation></semantics></math>
= number of samples</li>
<li>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>o</mi><annotation encoding="application/x-tex">o</annotation></semantics></math>
= number of objectives
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>o</mi><mo>=</mo><mn>2</mn></mrow><annotation encoding="application/x-tex">o=2</annotation></semantics></math>
in BioGA)</li>
<li>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>T</mi><annotation encoding="application/x-tex">T</annotation></semantics></math>
= number of generations</li>
</ul>
<div class="section level3">
<h3 id="component-complexities">Component Complexities<a class="anchor" aria-label="anchor" href="#component-complexities"></a>
</h3>
<table class="table">
<thead><tr class="header">
<th align="left">GA Component</th>
<th align="left">Complexity (Per Generation)</th>
<th align="left">Dominating Variables</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Initialization</td>
<td align="left"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>O</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>p</mi><mi>n</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">O(pn)</annotation></semantics></math></td>
<td align="left"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>p</mi><mo>,</mo><mi>n</mi></mrow><annotation encoding="application/x-tex">p, n</annotation></semantics></math></td>
</tr>
<tr class="even">
<td align="left">Fitness Evaluation</td>
<td align="left">
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>O</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>p</mi><mi>n</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>√ó</mo><mi>O</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>m</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">O(pn) \times O(m)</annotation></semantics></math><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mo>‚Üí</mo><annotation encoding="application/x-tex">\to</annotation></semantics></math><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>O</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>p</mi><mi>m</mi><mi>n</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">O(pmn)</annotation></semantics></math>
</td>
<td align="left"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>p</mi><mo>,</mo><mi>m</mi><mo>,</mo><mi>n</mi></mrow><annotation encoding="application/x-tex">p, m, n</annotation></semantics></math></td>
</tr>
<tr class="odd">
<td align="left">Selection (Sorting)</td>
<td align="left"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>O</mi><mrow><mo stretchy="true" form="prefix">(</mo><msup><mi>p</mi><mn>2</mn></msup><mi>o</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">O(p^2o)</annotation></semantics></math></td>
<td align="left">
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>
(quadratic)</td>
</tr>
<tr class="even">
<td align="left">Crossover (SBX)</td>
<td align="left"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>O</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>p</mi><mi>n</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">O(pn)</annotation></semantics></math></td>
<td align="left"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>p</mi><mo>,</mo><mi>n</mi></mrow><annotation encoding="application/x-tex">p, n</annotation></semantics></math></td>
</tr>
<tr class="odd">
<td align="left">Mutation (Adaptive)</td>
<td align="left"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>O</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>p</mi><mi>n</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">O(pn)</annotation></semantics></math></td>
<td align="left"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>p</mi><mo>,</mo><mi>n</mi></mrow><annotation encoding="application/x-tex">p, n</annotation></semantics></math></td>
</tr>
<tr class="even">
<td align="left">Replacement</td>
<td align="left"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>O</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>p</mi><mi>n</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">O(pn)</annotation></semantics></math></td>
<td align="left"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>p</mi><mo>,</mo><mi>n</mi></mrow><annotation encoding="application/x-tex">p, n</annotation></semantics></math></td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="total-complexity">Total Complexity<a class="anchor" aria-label="anchor" href="#total-complexity"></a>
</h3>
<p>The total complexity for
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>T</mi><annotation encoding="application/x-tex">T</annotation></semantics></math>
generations is dominated by the Fitness Evaluation and the Selection
phases:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mtext mathvariant="normal">Total Complexity</mtext><mo>=</mo><mi>O</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>T</mi><mo>‚ãÖ</mo><mrow><mo stretchy="true" form="prefix">(</mo><mi>p</mi><mi>m</mi><mi>n</mi><mo>+</mo><msup><mi>p</mi><mn>2</mn></msup><mi>o</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">
\text{Total Complexity} = O(T \cdot (pmn + p^2o))
</annotation></semantics></math> Factoring
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mtext mathvariant="normal">Total Complexity</mtext><mo>=</mo><mi>O</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>T</mi><mi>p</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>m</mi><mi>n</mi><mo>+</mo><mi>p</mi><mi>o</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">
\text{Total Complexity} = O(Tp(mn + po))
</annotation></semantics></math> This result highlights that scalability
is limited by the quadratic dependency on population size
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>
(due to sorting) and the linear dependency on the product of
genes/samples
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>m</mi><mi>n</mi></mrow><annotation encoding="application/x-tex">mn</annotation></semantics></math>
(due to fitness evaluation).</p>
</div>
</div>
<div class="section level2">
<h2 id="mathematical-optimization-interpretation">Mathematical Optimization Interpretation<a class="anchor" aria-label="anchor" href="#mathematical-optimization-interpretation"></a>
</h2>
<p>From a purely mathematical perspective, the BioGA algorithm serves as
a robust stochastic optimization solver for the following
multi-objective problem:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"><mrow><mtext mathvariant="normal">minimize </mtext><mspace width="0.333em"></mspace></mrow></mtd><mtd columnalign="left" style="text-align: left"><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>f</mi><mn>1</mn></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>P</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>,</mo><msub><mi>f</mi><mn>2</mn></msub><mrow><mo stretchy="true" form="prefix">(</mo><mi>P</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mrow><mtext mathvariant="normal">subject to </mtext><mspace width="0.333em"></mspace></mrow></mtd><mtd columnalign="left" style="text-align: left"><mi>P</mi><mo>‚àà</mo><msup><mi>‚Ñù</mi><mrow><mi>p</mi><mo>√ó</mo><mi>n</mi></mrow></msup></mtd></mtr></mtable><annotation encoding="application/x-tex">
\begin{aligned}
\text{minimize } &amp; (f_1(P), f_2(P)) \\
\text{subject to } &amp; P \in \mathbb{R}^{p \times n}
\end{aligned}
</annotation></semantics></math></p>
<p><strong>Core Characteristics that Favor GA:</strong></p>
<ol style="list-style-type: decimal">
<li>
<strong>Multi-Objective Nature:</strong> Explicitly seeking a set of
trade-off solutions (the Pareto front).</li>
<li>
<strong>High Dimensionality:</strong> Search space size is
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>p</mi><mo>√ó</mo><mi>n</mi></mrow><annotation encoding="application/x-tex">p \times n</annotation></semantics></math>,
making deterministic grid search intractable.</li>
<li>
<strong>Non-Differentiable Objectives:</strong> The sparsity
objective
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>f</mi><mn>2</mn></msub><annotation encoding="application/x-tex">f_2</annotation></semantics></math>
is non-differentiable, making standard calculus-based methods
impossible.</li>
<li>
<strong>Non-Convexity:</strong> The fitness landscape, when
combined, may be non-convex, meaning the GA‚Äôs global search capability
is essential to avoid local optima.</li>
</ol>
</div>
<div class="section level2">
<h2 id="biological-interpretation-and-relevance">Biological Interpretation and Relevance<a class="anchor" aria-label="anchor" href="#biological-interpretation-and-relevance"></a>
</h2>
<p>The mathematical operations within the BioGA framework have direct
analogues in the biological context of genomic data analysis, ensuring
the algorithm is biologically meaningful.</p>
<table class="table">
<colgroup>
<col width="33%">
<col width="33%">
<col width="33%">
</colgroup>
<thead><tr class="header">
<th align="left">Mathematical Component</th>
<th align="left">Biological Concept</th>
<th align="left">Function</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left"><strong>Population Initialization</strong></td>
<td align="left">Initial Biological Variability</td>
<td align="left">Starting the search process by sampling observed
expression levels.</td>
</tr>
<tr class="even">
<td align="left"><strong>Fitness
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>f</mi><mn>1</mn></msub><mo>,</mo><msub><mi>f</mi><mn>2</mn></msub></mrow><annotation encoding="application/x-tex">f_1, f_2</annotation></semantics></math>)</strong></td>
<td align="left">Functional Efficacy &amp; Parsimony</td>
<td align="left">
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>f</mi><mn>1</mn></msub><annotation encoding="application/x-tex">f_1</annotation></semantics></math>
ensures the optimized solution is relevant to observed data;
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>f</mi><mn>2</mn></msub><annotation encoding="application/x-tex">f_2</annotation></semantics></math>
enforces biological simplicity (Minimal Regulatory Model).</td>
</tr>
<tr class="odd">
<td align="left"><strong>Network Constraints</strong></td>
<td align="left">Gene-Gene Interactions</td>
<td align="left">Incorporating known biological constraints (e.g.,
regulatory or metabolic pathways) to guide mutation.</td>
</tr>
<tr class="even">
<td align="left"><strong>Clustering</strong></td>
<td align="left">Co-expressed Gene Modules</td>
<td align="left">Respecting known functional groupings during
initialization.</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>The <code>BioGA</code> package implements a theoretically sound and
computationally structured multi-objective evolutionary algorithm. Its
mathematical foundation‚Äîencompassing NSGA-II principles for
multi-objective handling, sophisticated SBX and adaptive
network-constrained mutation for exploration, and diversity-preserving
replacement‚Äîprovides a rigorous framework for simultaneously achieving
data fidelity and model sparsity in genomic data optimization.</p>
<details><summary><strong>Session Info</strong>
</summary><div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sessioninfo</span><span class="fu">::</span><span class="fu"><a href="https://sessioninfo.r-lib.org/reference/session_info.html" class="external-link">session_info</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">‚îÄ Session info ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ</span></span></span>
<span><span class="co">#&gt;  <span style="color: #555555; font-style: italic;">setting </span> <span style="color: #555555; font-style: italic;">value</span></span></span>
<span><span class="co">#&gt;  version  R version 4.5.2 (2025-10-31)</span></span>
<span><span class="co">#&gt;  os       Ubuntu 24.04.3 LTS</span></span>
<span><span class="co">#&gt;  system   x86_64, linux-gnu</span></span>
<span><span class="co">#&gt;  ui       X11</span></span>
<span><span class="co">#&gt;  language en</span></span>
<span><span class="co">#&gt;  collate  C.UTF-8</span></span>
<span><span class="co">#&gt;  ctype    C.UTF-8</span></span>
<span><span class="co">#&gt;  tz       UTC</span></span>
<span><span class="co">#&gt;  date     2025-12-04</span></span>
<span><span class="co">#&gt;  pandoc   3.1.11 @ /opt/hostedtoolcache/pandoc/3.1.11/x64/ (via rmarkdown)</span></span>
<span><span class="co">#&gt;  quarto   NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">‚îÄ Packages ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ</span></span></span>
<span><span class="co">#&gt;  <span style="color: #555555; font-style: italic;">package    </span> <span style="color: #555555; font-style: italic;">*</span> <span style="color: #555555; font-style: italic;">version</span> <span style="color: #555555; font-style: italic;">date (UTC)</span> <span style="color: #555555; font-style: italic;">lib</span> <span style="color: #555555; font-style: italic;">source</span></span></span>
<span><span class="co">#&gt;  bookdown      0.45    <span style="color: #555555;">2025-10-03</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  bslib         0.9.0   <span style="color: #555555;">2025-01-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  cachem        1.1.0   <span style="color: #555555;">2024-05-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  cli           3.6.5   <span style="color: #555555;">2025-04-23</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  desc          1.4.3   <span style="color: #555555;">2023-12-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  digest        0.6.39  <span style="color: #555555;">2025-11-19</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  evaluate      1.0.5   <span style="color: #555555;">2025-08-27</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  fastmap       1.2.0   <span style="color: #555555;">2024-05-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  fs            1.6.6   <span style="color: #555555;">2025-04-12</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  htmltools     0.5.8.1 <span style="color: #555555;">2024-04-04</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  htmlwidgets   1.6.4   <span style="color: #555555;">2023-12-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  jquerylib     0.1.4   <span style="color: #555555;">2021-04-26</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  jsonlite      2.0.0   <span style="color: #555555;">2025-03-27</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  knitr         1.50    <span style="color: #555555;">2025-03-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  lifecycle     1.0.4   <span style="color: #555555;">2023-11-07</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  pkgdown       2.2.0   <span style="color: #555555;">2025-11-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">any (@2.2.0)</span></span></span>
<span><span class="co">#&gt;  R6            2.6.1   <span style="color: #555555;">2025-02-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ragg          1.5.0   <span style="color: #555555;">2025-09-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  rlang         1.1.6   <span style="color: #555555;">2025-04-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  rmarkdown     2.30    <span style="color: #555555;">2025-09-28</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  sass          0.4.10  <span style="color: #555555;">2025-04-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  sessioninfo   1.2.3   <span style="color: #555555;">2025-02-05</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  systemfonts   1.3.1   <span style="color: #555555;">2025-10-01</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  textshaping   1.0.4   <span style="color: #555555;">2025-10-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  xfun          0.54    <span style="color: #555555;">2025-10-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  yaml          2.3.11  <span style="color: #555555;">2025-11-28</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [1] /home/runner/work/_temp/Library</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [2] /opt/R/4.5.2/lib/R/site-library</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [3] /opt/R/4.5.2/lib/R/library</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ</span></span></span></code></pre></div>
</details>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://danymukesha.github.io/" class="external-link">Dany Mukesha</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
